fix(source-klaviyo): sync all metric definitions regardless of configured start date

mkrawc opened the pull request:
## What

Klaviyo `metrics` table contains metric definitions that are referenced in other streams like events. In the events table, the `relationships.data.metric.id` value allows to join to the metrics table to expose the metric name. This is the only way, as I understand it, to know in the events table what event is being referenced. It is how we would understand if the event data is for an email received, opened, clicked, etc.

However when a start_date is configured for an Airbyte Klaviyo connector it fetches all  metrics that were updated since the configured start date. Because metrics usually don't change often most of metrics (in the sense of metric definitions) are not downloaded and there's no way to understand events data.

## How
In this pull request I change the initial start date for metrics table so that all metrics are initially downloaded regardless of the configured start_date.

## Review guide
As usual

## User Impact
For existing connections: none
For new connections: All metrics are saved regardless of when they were updated last.

## Can this PR be safely reverted and rolled back?
- [x] YES 💚
- [ ] NO ❌

mkrawc committed:
Fetch all klaviyo metrics

mkrawc committed:
Update klaviyo documentation

marcosmarxm reviewed (commented):
The change is needed because the start date is used for all streams and you want to keep using it for events which is a large stream compared to metrics. And metrics its important to capture them to join later with events and others streams.

mkrawc commented:
@marcosmarxm This is correct. Metrics is sort of metadata that is necessary to add context for other incremental streams like events. Think of it as metrics define types of events like email received, opened, clicked, and events return all occurrences when a user email is received, opened or clicked.

marcosmarxm reviewed (approved):
The change looks good and reasonable. We need final approval from the connector team and fix couple of integration tests are failing (not because your change)

mkrawc commented:
@marcosmarxm Any updates on this?

DanyloGL commented:
@mkrawc, could you please resolve conflicts? After this team will review your PR.
Thanks!

mkrawc commented:
@DanyloGL Rebased now.

DanyloGL committed:
Merge branch 'master' into klaviyo_all_metrics

DanyloGL committed:
Update metadata.yaml

DanyloGL committed:
Merge branch 'master' into klaviyo_all_metrics

DanyloGL committed:
Update metadata.yaml

DanyloGL committed:
Merge branch 'master' into klaviyo_all_metrics

darynaishchenko commented on airbyte-integrations/connectors/source-klaviyo/manifest.yaml:
```diff
@@ -772,7 +772,7 @@ definitions:
         cursor_field: updated
         datetime_format: "%Y-%m-%dT%H:%M:%S%z"
-        start_datetime: "{{ config.get('start_date', '2012-01-01T00:00:00Z') }}"
+        start_datetime: "{{ '1970-01-01T00:00:00Z' }}"
```
This change is applied for the pure stream also, not just for its child streams. I think we should keep its start date as is (e.g. what if someone wants to sync metrics starting from some date, not all). 
But implement one more metrics stream (call it all_metrics for example) with the updated start date that is only used for child streams in partition routers. 
We do't need to add this stream to list of available source stream, it's just an internal service stream.

Note: for all_metrics we can even remove incremental_sync definition since its client side incremental and api returns all metrics by default

bazarnov replied:
Refreshed this PR against current master and took a closer look at the suggested alternative:

- An internal metrics stream for the partition routers already exists — `metrics_for_reporting` (since 2.17.0) — and it is unaffected by `start_date`, so the reporting streams are already covered.
- The reported problem is about the records emitted by the user-facing `metrics` stream: with `start_date` configured (or the 1-year default since 2.19.0), older metric definitions never reach the destination, so `events.relationships.data.metric.id` cannot be joined to metric names. An internal-only stream cannot fix that — it never emits records.
- `GET /api/metrics` has no `updated` filter (only `integration.name`/`integration.category`), which is why this stream is client-side incremental: the full list is downloaded on every sync regardless of `start_date`, and the filter only drops already-downloaded records. Keeping the filter saves no API traffic — it only leaves the dimension table incomplete.

So the fix pins the metrics start to epoch while keeping the incremental cursor, meaning existing state/dedup behavior is unchanged. Also done on top: rebased onto current master (2.20.0), kept the 1Y-default guarantee for every other stream in `test_start_date_default.py`, added mock-server tests proving old definitions are emitted and that a saved state cursor still filters, and documented the exception in the spec and the docs page.

bazarnov committed:
Merge branch 'master' into klaviyo_all_metrics

bazarnov committed:
Apply metrics start-date fix on current master, add tests and docs

- pin the metrics stream start_datetime to epoch so the stream stays complete
  regardless of the configured start_date (the endpoint has no date filter;
  records are filtered client-side, so this costs no extra API traffic)
- bump to 2.20.0 on top of 2.19.3 and refresh the changelog entry
- exempt metrics in test_start_date_default.py while keeping the one-year
  default guarantee for every other stream
- add mock-server tests: definitions older than start_date are emitted, and
  an existing state cursor still filters old records
- document the metrics exception in the spec and the docs page

bazarnov reviewed (approved):
Verified against the Klaviyo API docs: `GET /api/metrics` has no date filter (only `integration.name`/`category`), the stream is client-side incremental — the full metrics list is downloaded on every sync regardless of `start_date`, so the old filter only dropped already-downloaded records and left the metrics dimension table incomplete (worse since 2.19.0, where the default start became one year back). Pinning the metrics start to epoch is the right fix and costs no extra API traffic; the incremental cursor and existing connection state are untouched.

Applied on top of the original change: refreshed on current master (2.20.0), kept the one-year-default guarantee for every other stream in `test_start_date_default.py`, added discriminating mock-server tests (definitions older than `start_date` are emitted; a saved state cursor still filters), and documented the metrics exception in the spec and docs. Full connector unit suite passes (141 tests).

octavia-squidington-iii committed:
chore: auto-fix lint and format issues
